soc/arm/nxp_kinetis/ke1xf/soc.c

```diff
@@ -272,46 +272,24 @@ static int ke1xf_init(const struct device *arg)
 	 */
 	NMI_INIT();
 
-#ifdef CONFIG_KINETIS_KE1XF_ENABLE_CODE_CACHE
-	L1CACHE_EnableCodeCache();
+#ifndef CONFIG_KINETIS_KE1XF_ENABLE_CODE_CACHE
+	/* SystemInit will have enabled the code cache. Disable it here */
+	L1CACHE_DisableCodeCache();
 #endif
 	/* Restore interrupt state */
 	irq_unlock(old_level);
 
 	return 0;
 }
 
-void z_arm_watchdog_init(void)
-{
-	/*
-	 * NOTE: DO NOT SINGLE STEP THROUGH THIS FUNCTION!!! Watchdog
-	 * reconfiguration must take place within 128 bus clocks from
-	 * unlocking. Single stepping through the code will cause the
-	 * watchdog to close the unlock window again.
-	 */
+#ifdef CONFIG_PLATFORM_SPECIFIC_INIT
 
-	/* Unlock watchdog to enable reconfiguration after bootloader */
-	WDOG->CNT = WDOG_UPDATE_KEY;
-	while (!(WDOG->CS & WDOG_CS_ULK_MASK)) {
-		;
-	}
-
-	/*
-	 * Watchdog reconfiguration only takes effect after writing to
-	 * both TOVAL and CS registers.
-	 */
-#ifdef CONFIG_WDOG_ENABLE_AT_BOOT
-	WDOG->TOVAL = CONFIG_WDOG_INITIAL_TIMEOUT >> 1;
-	WDOG->CS = WDOG_CS_PRES(1) | WDOG_CS_CLK(1) | WDOG_CS_WAIT(1) |
-		   WDOG_CS_EN(1) | WDOG_CS_UPDATE(1);
-#else /* !CONFIG_WDOG_ENABLE_AT_BOOT */
-	WDOG->TOVAL = 1024;
-	WDOG->CS = WDOG_CS_EN(0) | WDOG_CS_UPDATE(1);
-#endif /* !CONFIG_WDOG_ENABLE_AT_BOOT */
-
-	while (!(WDOG->CS & WDOG_CS_RCS_MASK)) {
-		;
-	}
```

Comment:
This breaks the current behaviour where Kconfig allows enabling the watchdog at boot with an initial timeout. We rely on this as we use it as part of our MCUboot update strategy.

Reply:
thank you for highlighting this. I've added the watchdog configuration back for the KE18F, and now only reconfigure the watchdog if `CONFIG_WDOG_ENABLE_AT_BOOT=y`, since otherwise the watchdog would be disabled at boot.

Reply:
Will this still leave a window where the watchdog is disabled (`SystemInit()` will disable it, `z_arm_watchdog_init()` will re-enable it)?

Reply:
Yes, it will. I suppose this differs from the current behavior, where the watchdog is only reconfigured at boot if this KConfig is enabled.

NXP would like to convert all our SOCs to use SystemInit (so that we will include any fixes for errata or other important boot configuration that the SDK may be doing and we need to be aware of), but I'm aware that the KE1X is a bit of a special case.

I supposed one option we could use for the KE1x is to `imply PLATFORM_SPECIFIC_INIT` rather than `select` it, so that your downstream implementation can skip `SystemInit`. If I dropped the `z_arm_watchdog_init` code, that could be a hook point for your application to manage the watchdog and preserve your MCUBoot strategy. Does that seem like a good solution to you?

Reply:
Hmm. I'd much rather see NXP carrying their hardware policy of leaving the watchdog enabled at boot into Zephyr, with an option to turn it off - or some variant of that.

Having the watchdog enabled at boot is the safest way to ensure that a failed update will automatically roll back to the previous firmware. Having the default NXP behaviour in Zephyr be "boot" -> "disable hardware enabled watchdog" -> "re-enable hardware watchdog" kind of goes against one of the best features of the NXP watchdog hardware.

Why does `SystemInit()` disable the hardware watchdog by default? Could this be changed to be optional, so that we can control it from a Kconfig?